src/gpt.py

```diff
@@ -0,0 +1,312 @@
+import asyncio
+import datetime
+import json
+import random
+import re
+from typing import Dict, List, Optional, Tuple
+
+import openai
+import tiktoken
+from nextcord import Member, Message
+
+import config
+
+
+# Initialize globals
+def initialize(client):
+    global gpt_lock
+    global context_data
+    global encoding
+    global self_user
+    global banned_word_pattern
+    global word_trigger_pattern
+    global user_trigger_pattern
+    global user_info_map
+
+    openai.api_key = config.openai_api_key
+    # Global lock for message response
+    gpt_lock = asyncio.Lock()
+    # Load manual hidden data
+    try:
+        with open(config.gpt_context_file) as f:
+            context_data = json.load(f)
+    except (FileNotFoundError, json.decoder.JSONDecodeError) as e:
+        print(
+            f"ERROR: Issue loading {config.gpt_context_file}. GPT capabilities will be disabled.\n{e}"
+        )
+        # Signal context data was not loaded correctly
+        context_data = None
+        return
+    # Preload tokenizer
+    encoding = tiktoken.encoding_for_model(config.OPENAI_MODEL)
+    # Store bot user
+    self_user = client.user
+    # Cache regex for banned words
+    banned_word_pattern = re.compile(
+        r"\b{}\b".format("|".join(context_data["banned_phrases"]))
+    )
+
+    # Cache regex for word triggers
+    word_trigger_pattern = re.compile(
+        r"\b({})\b".format("|".join(context_data["word_triggers"].keys()))
+    )
+
+    # Cache regex for user info triggers
+    user_trigger_pattern = re.compile(
+        r"\b({})\b".format(
+            "|".join(
+                [
+                    user["name"].lower()
+                    for user in context_data["user_info"].values()
+                    if "info" in user
+                ]
+            )
+        )
+    )
+    # Store map for user info triggers
+    user_info_map = {
+        user["name"].lower(): user["info"].lower()
+        for user in context_data["user_info"].values()
+        if ("info" in user and "name" in user)
+    }
+
+
+# Check if a message should not be allow in API history
+# Currently this is just if it contains banned phrases
+def disallowed_message(message: Message) -> bool:
+    return banned_word_pattern.search(message.content.lower()) is not None
+
+
+# Get the name we should call the user
+def get_name(user):
+    user_info = context_data["user_info"]
+    id = str(user.id)
+    if id in user_info and "name" in user_info[id]:
+        return user_info[id]["name"]
+    return user.name
```

Comment:
I chose Discord name instead of nick because people seem to have constantly changing nicknames in the server that we don't call them. Just now kley is "kylie", fin is "“Real” Russian Bank Assistant", ember is "embird", emi is "emillion", kat is "kat burglar", etc. Yet in convos we still tend to refer to them in convo by their "true" name. Some of this is fixed by the hardcoded nicknames list but I think there's a much higher "true name" hit rate by nick and not by whatever joke they they currently have.

Reply:
That's fair. On the contrary Evershade's discord username is "sillyzippy" which typically nobody calls them. And it may miss out on fun opportunities for people to change their name to "Obama" or whatnot and start talking with the bot. But it'll have to be something we gauge in practice I think.

For now let's leave it as it is. We should probably strip out the `#1234` that is appended to the end of Discord usernames though?

Reply:
> We should probably strip out the #1234 that is appended to the end of Discord usernames though?

`user.name` already has the discriminator stripped